Classes/Hooks/HookProvider.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OliverKlee\Seminars\Hooks;
+
+use OliverKlee\Seminars\Hooks\Interfaces\Hook;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+/**
+ * Provides functions for unified hooks
+ *
+ * A hook allows adding functionality at certain points of the program path. These points are
+ * grouped using an interface, declaring the methods to implement to hook in.
+ *
+ * Hooking in
+ * The FQCN of the interface is the key to `$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']`
+ * where you register your hooked-in classes. Your class will be instantiated once when the first
+ * point is reached and re-used for all other points.
+ *
+ * Implementing hook points
+ * Instantiate this class with the interface you need implemented. First call to `executeHook()` will
+ * instantiate the registered classes. Every further call will reuse the same instances. On each
+ * call provide the method required at the point in your program.
+ *
+ * There is an optional index to `$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']`, provided
+ * for easier conversion of existing hooks to this class.
+ *
+ * @author Michael Kramer <m.kramer@mxp.de>
+ */
+class HookProvider
+{
+    /**
+     * @var string
+     */
+    protected $interfaceName = '';
+
+    /**
+     * Index in `$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']` of hooked-in classes
+     *
+     * @var string
+     */
+    protected $index = '';
+
+    /**
+     * @var Hook[]
+     */
+    protected $hookObjects = [];
+
+    /**
+     * @var bool
+     */
+    protected $hooksHaveBeenRetrieved = false;
```

Comment:
Generally, I prefer private fields and methods for things that are not intented to be used in subclasses/Xclasses. Could you please check the fields and methods of this class for this? Thanks!

Reply:
Generally, as I often want to re-use code by extending other projects classes, I learned to hate things being unneccesarily private. In fact, I regularly end up copying large code portions into subclasses (that can never ever be thought of by the original author) because of one single unnecessarily private field.

I will, of course, make everything as `private` as possible if You want me to. But why urge someone wanting to extend it to re-implement it completely?

IMHO, `protected` is **the** way to state: This is internal. You may use it, but don't rely on it being the same in next version (the opposite of API / `public`). And `private` says: I have security concerns or other very strong reasons so I need to keep this out of anybodies reach.

Reply:
It's mostly about XCLASSes and refactoring: With TYPO3 and its XCLASS mechanism, people tend to rely on the protected methods staying there and working the same. Refactoring those protected methods and fields (renaming, removing …) breaks those XCLASS usages. That's why I'd like to reduce the surface that XCLASSes can directly use.

And for the code to be maintainable, I need to be able to refactor it fearlessly. In addition to that, static code analysis can detect unused methods and fields a lot easier and safer if they are private, not protected (or public).

Reply:
(In other words, I have a different understanding than you what "protected" means in the TYPO3 world.)

Reply:
I am one of those XCLASS users, and very well know what you intend. But: You **should** not care for those that XCLASS Your code when thinking of refactoring. That is, what the XCLASSer is risking when XCLASSing. If it would be possible to do what needs to be done using Observers, Hooks and similar, this would be done. For sure.

XCLASSing (and even in project re-using) is impossible, if there is only ONE field involved that is `private`. Simple example: I tried to XCLASS Your `DefaultController` to add another hook. I stopped after copying most methods into my subclass, and instead changed the extension itself to include my custom hook. Thanks to git this works well even on larger refactorings. But it **should** not be done that way, I should have an extension XCLASSing DefaultController.

Reply:
Okay, then let's keep them protected.